### DISCUSSION TITLE
ESP32-Cam: stuttering

### DISCUSSION THREAD

**Discussion by bartplessers** [arendst/Tasmota#13788] [Q&A] [answered]
Hi everyone

I have some experience with tasmota on sonoff devices, but this is my first esp32-cam project.
So apologies for my ignorance...

I managed to flash tasmota 10.0.0 on a "ESP32-CAM WiFi + bluetooth Camera Module Development Board ESP32 With Camera Module OV2640"
https://www.banggood.com/ESP32-CAM-WiFi-+-bluetooth-Camera-Module-Development-Board-ESP32-With-Camera-Module-OV2640-p-1394679.html

Then
- configure module: "Module type (ESP32-DevKit)" --> "ESP32-Cam (2)"
- console: 'WcResolution	13"
- console: "RTSP init"

I can connect to my device:
The MJPEG stream is accessible at 
http://192.168.3.121:81/stream
 or 
http://192.168.3.121:81/cam.mjpeg

A single snapshot can be obtained at 
http://192.168.3.121:80/snapshot.jpg

But I noticed that the stream is "stuttering", very low fps.
Is this normal?
I know this is extremely cheap software, so I may not expect a lot of it, but I am a little disappointed in that low framerate.
But maybe there is something I need to config?

I read somewhere maybe I should change my "template":
web ui --> Configure other --> Other parameters --> Template "xxxxxxxx"
but don't now what I have to use for my device.

Or ii this just the normal behaviour for this kind of device?

any help is appreciated.

kind regards,
Bart Plessers

**Comment by ascillato** [comment, marked as answer]
> I know this is extremely cheap software, so I may not expect a lot of it

???? what do you mean? Tasmota Software or the ESP32Cam Hardware?

> But maybe there is something I need to config?

Maybe. Mine works really good. No stuttering.

* Have you used the ESP32CAM Tasmota precompiled version which uses 240MHz? Or have you compiled by yourself?
* Have you applied the template https://templates.blakadder.com/ai-thinker_ESP32-CAM.html ?
* Do you have a suitable power source? ESP32Cam uses more energy than a regular ESP32.
* Have you tried latest Tasmota from development branch? ( http://ota.tasmota.com/tasmota32/tasmota32-webcam.bin )
* Have you tried changing resolution with the following commands?
https://github.com/arendst/Tasmota/blob/50ec8a541e736892ee55fdfea5fc912014bee091/tasmota/xdrv_81_esp32_webcam.ino#L22-L75

**Reply by sfromis** [reply]
When you step up to `WcResolution 13`, it is **entirely normal** with a rather low frame rate. If you want a webcam, buy one, and pay for it. ESP32-CAM is a cheap device, do not expect the same as when you pay 10 times as much. For tolerable frame rates, lower the resolution.

**Reply by jenniferlee1818** [reply]
@ascillato @sfromis Which is the correct doc for the WcResolution command?  
this one: [xdrv_81_esp32_webcam.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_81_esp32_webcam.ino)
or this one: [https://tasmota.github.io/docs/Commands/#camera](https://tasmota.github.io/docs/Commands/#camera)

**Reply by sfromis** [reply]
Maybe neither? When I set `WcResolution 10`, I get a frame width of 1140 and height 856 (approx), not 1024x768 or 1600x1200, or even anything appearing on the lists.

**Reply by ascillato** [reply]
> Which is the correct doc for the WcResolution command?

With my device (espcam ai-thinker) the resolutions explained in the driver are correct.

The docs are not matching the ones of the driver.

> When I set WcResolution 10, I get a frame width of 1140 and height 856 (approx), not 1024x768 or 1600x1200, or even anything appearing on the lists.

So, it may depend on the camera type (?)

**Reply by ascillato** [reply]
At least, maybe we should make the docs match the resolutions explained in the driver I think.

**Reply by sfromis** [reply]
While I measured the pixels mentioned above with a screen ruler, afterwards using Chrome tools to find resolution got me to 1024x768 instead of what I saw on my monitor, even though I was careful not to scale content up/down from the default. Looks like there may be some extra level of scaling involved. Thus it did not appear to depend on camera type, but on what's used to view the stream.

**Reply by sfromis** [reply]
The driver was changed 13 months ago.... https://github.com/arendst/Tasmota/pull/11035
And https://github.com/arendst/Tasmota/issues/11009 says that
> the driver has been updated recently and new settings have been inserted which moved the UXGA setting to position 13!

Docs was not changed to match.

**Comment by bartplessers** [comment]
>???? what do you mean? Tasmota Software or the ESP32Cam Hardware?

:-)

>Have you used the ESP32CAM Tasmota precompiled version which uses 240MHz? Or have you compiled by yourself?

I followed instructions from https://cgomesu.com/blog/Esp32cam-tasmota-webcam-server/ so I downloaded
`https://ota.tasmota.com/tasmota32/release/tasmota32-webcam.bin`


>Have you applied the template https://templates.blakadder.com/ai-thinker_ESP32-CAM.html ?

yes


>Do you have a suitable power source? ESP32Cam uses more energy than a regular ESP32.

yes


>Have you tried latest Tasmota from development branch? ( http://ota.tasmota.com/tasmota32/tasmota32-webcam.bin )

no, I used the release. I will try with the dev branch and let you know if this makes any difference


>Have you tried changing resolution with the following commands?

yes, I'm running now with
WcResolution 9

but I have the feeling that the resolution is not much a problem. It's just that there is a latency of +/- 3 seconds and "fast" moving objects are more displayed as snapshots with 1 sec between. So it feels more like a low frames per second rate....


thanx for support!

**Reply by sfromis** [reply]
Low frames per second is a very different thing from latency. With `WcResolution 9`, I'm seeing bearable but modest FPS, with "several" frames per second. Latency from my finger moving in front of the lens till seeing it on my monitor via the Tasmota web UI is percievable but fractions of a second.

Important performance factors can be having a good power supply, and sufficient cooling, without that it gets to be a mess. Answering "yes" to "suitable" is no info at all.

**Reply by bartplessers** [reply]
>Low frames per second is a very different thing from latency

The original question was "But I noticed that the stream is stuttering, very low fps. Is this normal?"



>Answering "yes" to "suitable" is no info at all.

Anchor USB power 63W. If I answer a "yes" its a yes. Even if the question was "suitable"

**Reply by sfromis** [reply]
The watts rating of your USB wart does still not tell how you power the camera board, which is not USB powered. If you have the "double-decker" with ESP32-CAM-MB, that piece of hardware seems to be of rather low quality, meaning that my confidence in the power delivery would still be low.

And the question about cooling? Allowing the board to heat too much can be very detrimental to performance.

Your varying wording about "stuttering" has made it unclear how you use the words. Therefore I provided a more precise explanation of what I experience, not for the purpose of you just quoting yourself, but as a reference for expectations when doing what the board should be able to.

Your initial "very low fps" was most likely due to unrealistic expectations about using a high resolution, where a very low frame rate is actually normal, as stated already. Going down to WcResolution 9 instead of 13 helps a lot when I try.

**Comment by ossapandra** [comment]
I got very bad framerate/stuttering too and I assumed the ESP could not handle it performance wise. But it turned out to be the low WIFI signal strength. So my advice would be to test your setup right beside your router.
